app/notification_test.go

```diff
@@ -2410,3 +2410,106 @@ func TestInsertGroupMentions(t *testing.T) {
+		require.Len(t, groupsMap, 0)
+	})
+
+	group1.AllowReference = true
+	group1, err = th.App.UpdateGroup(group1)
+	require.Nil(t, err)
+
+	group2 := th.CreateGroup()
+	t.Run("should only return groups with allow reference", func(t *testing.T) {
+		groupsMap, err = th.App.getGroupsAllowedForReferenceInChannel(channel, team)
+		require.Nil(t, err)
+		require.Len(t, groupsMap, 1)
+		require.Nil(t, groupsMap[group2.Name])
+		require.Equal(t, groupsMap[group1.Name], group1)
+	})
+
+	group2.AllowReference = true
+	group2, err = th.App.UpdateGroup(group2)
+	require.Nil(t, err)
+
+	// Sync first group to constrained channel
+	constrainedChannel := th.CreateChannel(th.BasicTeam)
+	constrainedChannel.GroupConstrained = model.NewBool(true)
+	constrainedChannel, err = th.App.UpdateChannel(constrainedChannel)
+	require.Nil(t, err)
+	_, err = th.App.UpsertGroupSyncable(&model.GroupSyncable{
+		GroupId:    group1.Id,
+		Type:       model.GroupSyncableTypeChannel,
+		SyncableId: constrainedChannel.Id,
+	})
+	require.Nil(t, err)
+
+	t.Run("should return only groups synced to channel if channel is group constrained", func(t *testing.T) {
+		groupsMap, err = th.App.getGroupsAllowedForReferenceInChannel(constrainedChannel, team)
+		require.Nil(t, err)
+		require.Len(t, groupsMap, 1)
+		require.Nil(t, groupsMap[group2.Name])
+		require.Equal(t, groupsMap[group1.Name], group1)
+	})
+
+	// Create a third group not synced with a team or channel
+	group3 := th.CreateGroup()
+	group3.AllowReference = true
+	group3, err = th.App.UpdateGroup(group3)
+	require.Nil(t, err)
+
+	// Sync group2 to the team
+	team.GroupConstrained = model.NewBool(true)
+	team, err = th.App.UpdateTeam(team)
+	require.Nil(t, err)
+	_, err = th.App.UpsertGroupSyncable(&model.GroupSyncable{
+		GroupId:    group2.Id,
+		Type:       model.GroupSyncableTypeTeam,
+		SyncableId: team.Id,
+	})
+	require.Nil(t, err)
+
+	t.Run("should return union of groups synced to team and any channels if team is group constrained", func(t *testing.T) {
+		groupsMap, err = th.App.getGroupsAllowedForReferenceInChannel(channel, team)
+		require.Nil(t, err)
+		require.Len(t, groupsMap, 2)
+		require.Nil(t, groupsMap[group3.Name])
+		require.Equal(t, groupsMap[group2.Name], group2)
+		require.Equal(t, groupsMap[group1.Name], group1)
+	})
+
+	t.Run("should return only subset of groups synced to channel for group constrained channel when team is also group constrained", func(t *testing.T) {
+		groupsMap, err = th.App.getGroupsAllowedForReferenceInChannel(constrainedChannel, team)
+		require.Nil(t, err)
+		require.Len(t, groupsMap, 1)
+		require.Nil(t, groupsMap[group3.Name])
+		require.Nil(t, groupsMap[group2.Name])
+		require.Equal(t, groupsMap[group1.Name], group1)
+	})
+
+	team.GroupConstrained = model.NewBool(false)
+	team, err = th.App.UpdateTeam(team)
+	require.Nil(t, err)
+
+	t.Run("should return all groups when team is not group constrained but channel that is group constrained exists on the team", func(t *testing.T) {
```

Comment:
why the "but channel that is group constrained exists on the team" ? How is it different from the second test case "should return only groups synced to channel if channel is group constrained" ?

Reply:
In this test case there is a channel on the team that is group constrained, but the channel that we're passing as a parameter is different from the group constrained one

Reply:
Do you think this test is not needed?

Reply:
I guess the question is: does the "when team is not group constrained but channel that is group constrained exists on the team" matters in this case - or maybe we can rephrase as - "current channel is not group mentioned and the team it is part of is not group constrained either". Not clear if the the fact that there might be one or more channels group-constrained in the team matters.

Reply:
Done @catalintomai I have made the test description clearer